### PR TITLE
[`pylint`] Implement `unnecessary-dict-index-lookup` (`PLR1733`)

### PR DESCRIPTION
**Pull request by diceroll123**
## Summary

Add [R1733](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/unnecessary-dict-index-lookup.html) and autofix!

See #970 

## Test Plan

`cargo test` and manually

### CODE REVIEW

**Comment by diceroll123** [comment]
This PR is practically just #7999 in a trenchcoat

**Comment by Skylion007** [comment]
Does this pylint rule cover dict comprehensions? or would that be covered by another rule? Or is it just not covered sadly?

**Comment by Skylion007** [comment]
@charliermarsh Anything blocking these two PRs btw?

**Comment by charliermarsh** [comment]
@Skylion007 - Just me finding time to review it, there's a lot of logic so I just need to read it closely.

**Comment by Skylion007** [comment]
I would add a test case for this bug: https://github.com/astral-sh/ruff/pull/7999#issuecomment-1793503531 . I suspect you would hit it with dict comprehensions as well.

File: crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_dict_index_lookup.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    {FRUITS[fruit_name] for fruit_name, _ in FRUITS.items()}  # PLR1733
+    {fruit_name: FRUITS[fruit_name] for fruit_name, _ in FRUITS.items()}  # PLR1733
+
+    for fruit_name, _ in FRUITS.items():
+        print(FRUITS[fruit_name])  # Ok
+        blah = FRUITS[fruit_name]  # Ok
+        FRUITS[fruit_name]: int = FRUITS[fruit_name]  # Ok
+        FRUITS[fruit_name] = FRUITS[fruit_name]  # Ok
+        FRUITS[fruit_name] += FRUITS[fruit_name]  # Ok
+        FRUITS[fruit_name] = "d"  # Ok
```
Comment (Skylion007) [inline crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_dict_index_lookup.py:32]: What about accessing `_` after `FRUITS[fruit_name]` is edited? that's the main thing missing here.

File: crates/ruff_linter/src/codes.rs [hunk 2]
```diff
@@ -254,6 +254,7 @@ pub fn code_to_rule(linter: Linter, code: &str) -> Option<(RuleGroup, Rule)> {
         (Pylint, "R1711") => (RuleGroup::Stable, rules::pylint::rules::UselessReturn),
         (Pylint, "R1714") => (RuleGroup::Stable, rules::pylint::rules::RepeatedEqualityComparison),
         (Pylint, "R1706") => (RuleGroup::Preview, rules::pylint::rules::AndOrTernary),
+        (Pylint, "R1733") => (RuleGroup::Preview, rules::pylint::rules::UnnecessaryDictIndexLookup),
```
Comment (Skylion007) [inline crates/ruff_linter/src/codes.rs:257]: Shouldn't this be 1 row down to maintain sorted order?

Comment (diceroll123) [reply crates/ruff_linter/src/codes.rs]: rebased and fixed!

**Review by charliermarsh — COMMENTED** [review]
@diceroll123 -- Thank you! Do you mind taking a look at some of the tweaks I made in https://github.com/astral-sh/ruff/pull/8932 and the comments I left to explain them, and applying those changes here?

**Comment by diceroll123** [comment]
Will do, I see you're still adding comments so I'll check back later!

**Review by charliermarsh — APPROVED** [review]
Thanks!

**Comment by diceroll123** [comment]
Thank YOU! 😄

**Pull request merged**